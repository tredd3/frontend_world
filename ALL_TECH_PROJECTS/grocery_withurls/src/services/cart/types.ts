import { actions } from './shared';

export type NonCheckoutParams = { action: typeof actions.preCheckout };
export type CheckoutParams = {
  action: typeof actions.postCheckout;
  // customerAddressId is actually a string in the request. Need to check if number works.
  // If not, we'll have to cast before sending it.
  customerAddressId: number;
  deliveryDate: string;
  timeslot: '';
  isAutoSelected: 0;
};

export type MaybeCheckout = CheckoutParams | NonCheckoutParams;

export type AddCartItemRequestBase = {
  userId: number;
  storeId?: number;
  cartId: number;
  product: { skuId: number; qty: number; productType: number }[];
};

export type AddCartItemRequest = AddCartItemRequestBase & NonCheckoutParams;

export type ChangeQuantityRequest = AddCartItemRequestBase & MaybeCheckout;

export type RemoveCartItemRequest = {
  userId: number;
  storeId?: number;
  cartId: number;
  skuId: number;
  qty: 0;
  type: number;
} & MaybeCheckout;

export type MoveCartItemRequest = {
  userId: number;
  cartId: number;
  skuId: number;
  qty: number;
  productType: number;
  moveType: 'SaveForLater' | 'MoveToCart';
};

export type DeleteSavedForLaterItemRequest = {
  skuId: number;
  qty: number;
  productType: number;
  productCategoryType: 'SaveForLater';
};

export type PaymentRequestRequest = {
  Amount: number;
  CartId: number;
  MobileNo: number;
  StoreId: number;
  UserId: number;
};

export type PaymentRequestResponse = {
  SUPERMID: string;
  CLIENTID: string;
  MERCHANTNAME: string;
  SPLITTXNFLAG: string;
  CHANNEL: string;
  CURRENCY: string;
  TXNTYPE: string;
  Time: string;
  Checksum: string;
  Amount: string;
  RefNo: string;
  SplitDetails: string;
  ReturnUrl: string;
  PaymentUrl: string;
  Version: string;
  IncludePayModes: string;
  ExcludePayModes: string;
  ProductDescription: string;
  UDF1: string;
  UDF2: string;
  UDF3: string;
  UDF4: string;
  UDF5: string;
};

export type PGPostPayload = {
  merchantid: string;
  clientid: string;
  merchantname: string;
  splittxnflag: string;
  channel: string;
  'transaction.currency': string;
  'transaction.txntype': string;
  'transaction.timestamp': string;
  checksum: string;
  'transaction.amount': string;
  'transaction.extref': string;
  'transaction.splitdetails': string;
  returl: string;
  version: string;
  'preferences.includepaymodes': string;
  'preferences.excludepaymodes': string;
  productdescription: string;
  udf1: string;
  udf2: string;
  udf3: string;
  udf4: string;
  udf5: string;
  token: string;
  'subscriber.customername': string;
  'subscriber.mobilenumber': string;
  'subscriber.customerid': string;
}

export type DeliveryModesAPIRequestType = {
  StoreId: number;
  OrderValue: number;
  Pincode: number;
};

export type CheckoutRequest = {
  storeId: number;
  cartId: number;
  userId: number;
  // Not sure if the user's mobile number is needed
  // userMobile: number;
} & CheckoutParams;

export type PersistedPaymentDetails = {
  refNo: string;
  userId: number;
  storeId: number;
  cartId: number;
};

export type PaymentResponseRequest = {
  RefNo: string;
  UserId: number;
  StoreId: number;
};

export type PlaceCODOrderRequest = {
  CartId: number;
  StoreId: number;
  PaymentMethod: 1;
  MID: 0;
}
