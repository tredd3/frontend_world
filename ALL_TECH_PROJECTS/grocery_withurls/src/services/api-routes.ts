const baseUrl = process.env.REACT_APP_BASE_URL || '';
const baseUrlLogin = process.env.REACT_APP_BASE_URL_LOGIN || '';

export const apiRoutes = {
  getNewAccessToken: `${baseUrl}/jm/auth/oauth/v2/token`,
  getDashboard: `${baseUrl}/jiokart/v1/jiogrocery/get/dashboard`,
  viewAllDashboard: `${baseUrl}/jiokart/v1/jiogrocery/view/all/dashboard`,
  getStores: `${baseUrl}/jiokart/v1/get/stores`,
  getCart: `${baseUrl}/jiokart/v1/jiogrocery/get/cart`,
  addCartItem: `${baseUrl}/jiokart/v1/add/cart/item`,
  removeCartItem: `${baseUrl}/jiokart/v1/remove/cart/item`,
  moveCartItem: `${baseUrl}/jiokart/v1/jiogrocery/move/item`,
  getCategories: `${baseUrl}/jiokart/v1/get/categories`,
  getProductList: `${baseUrl}/jiokart/v1/jiogrocery/store/product`,
  checkServiceability: `${baseUrl}/jiokart/v1/jiogrocery/pincode/check`,
  getProductSearch: `${baseUrl}/jiokart/v1/jiogrocery/product/search`,
  getSuggestions: `${baseUrl}/jiokart/v1/jiogrocery/suggestion`,
  addUser: `${baseUrl}/jiokart/v1/jiogrocery/add/user`,
  saveAddress: `${baseUrl}/jiokart/v1/jiogrocery/save/user/address`,
  getAddress: `${baseUrl}/jiokart/v1/jiogrocery/get/user/address`,
  deleteAddress: `${baseUrl}/jiokart/v1/jiogrocery/remove/user/address`,
  getProductDetails: `${baseUrl}/jiokart/v1/jiogrocery/store/product`,
  getSimilarItems: `${baseUrl}/jiokart/v1/jiogrocery/store/similar/items`,
  getOrderDetails: `${baseUrl}/jiokart/v1/jiogrocery/order/details`,
  getOrderHistory: `${baseUrl}/jiokart/v1/jiogrocery/order/listing`,
  getDeliveryModes: `${baseUrl}/jiokart/v1/get/delivery/modes`,
  getLocation: `${baseUrl}/jiokart/v1/jiogrocery/google/places`,
  searchOrderHistory: `${baseUrl}/jiokart/v1/jiogrocery/order/item/search`,
  checkout: `${baseUrl}/jiokart/v1/order/checkout`,
  placeOrder: `${baseUrl}/jiokart/v1/place/order`,
  getUserDetail: `${baseUrl}/jiokart/v1/jiogrocery/get/user`,
  getWishListData: `${baseUrl}/jiokart/v1/jiogrocery/get/item/list`,
  addItemToWishList: `${baseUrl}/jiokart/v1/jiogrocery/add/item`,
  removeItemFromWishList: `${baseUrl}/jiokart/v1/jiogrocery/remove/item`,
  sendOtp: `${baseUrlLogin}/idam/trialwifi/v3/dip/otp/send`,
  verifyOtp: `${baseUrlLogin}/idam/trialwifi/v3/dip/otp/verify`,
  validateuser: `${baseUrl}/cr/v1/validateuser`,
  cancelOrder: `${baseUrl}/jiokart/v1/order/cancel`,
  config: `${baseUrl}/jiokart/v1/logged/customer/configurations`,
  updateUser: `${baseUrl}/jiokart/v1/jiogrocery/update/user`,
  orderRating: `${baseUrl}/jiokart/v1/order/rating`,
  getPaymentRequest: `${baseUrl}/jiokart/v1/get/payment/request`,
  getPaymentResponse: `${baseUrl}/jiokart/v1/get/payment/response`,
  getNotifications: `${baseUrl}/jiokart/v1/jiogrocery/notification/get`,
  saveNotification: `${baseUrl}/jiokart/v1/jiogrocery/notification/save`,
  clearNotifications: `${baseUrl}/jiokart/v1/jiogrocery/clear/notification`,
  removeCartSavedItem: `${baseUrl}/jiokart/v1/jiogrocery/remove/item`
} as const;

const endpointsWithoutAuth = [
  apiRoutes.checkServiceability,
  apiRoutes.getStores,
  apiRoutes.getProductDetails,
  apiRoutes.getSimilarItems,
  apiRoutes.getLocation,
  apiRoutes.sendOtp,
  apiRoutes.getDashboard
];

const endpointsNeedingEnvelope = [
  apiRoutes.getCategories,
  apiRoutes.addCartItem,
  apiRoutes.removeCartItem,
  apiRoutes.getStores,
  apiRoutes.getDeliveryModes,
  apiRoutes.checkout,
  apiRoutes.placeOrder,
  apiRoutes.getPaymentRequest,
  apiRoutes.getPaymentResponse,
  apiRoutes.orderRating
];

const endpointsNeedingStoreId = [
  apiRoutes.getCart,
  apiRoutes.addCartItem,
  apiRoutes.removeCartItem,
  apiRoutes.placeOrder,
  apiRoutes.getDeliveryModes,
  apiRoutes.checkout,
  apiRoutes.placeOrder,
  apiRoutes.saveAddress,
  apiRoutes.getAddress,
  apiRoutes.getPaymentResponse,
  apiRoutes.removeCartSavedItem
];

export const needsEnvelope = (url: string) => endpointsNeedingEnvelope.includes(url);
export const needsAuthentication = (url: string) => !endpointsWithoutAuth.includes(url);
export const needsStoreId = (url: string) => endpointsNeedingStoreId.includes(url);

export const v2ApiRoutes = [
  apiRoutes.getPaymentResponse,
  apiRoutes.getOrderHistory,
  apiRoutes.getOrderDetails,
  apiRoutes.getCart,
  apiRoutes.getDeliveryModes
];
