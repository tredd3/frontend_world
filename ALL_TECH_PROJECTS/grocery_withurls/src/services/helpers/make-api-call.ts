/* eslint-disable no-console */
import uuidV1 from 'uuid/v1';
import { needsEnvelope, needsAuthentication, v2ApiRoutes } from '../api-routes';
import { getTokens } from '../../intents';
import { HTTPError, APIError } from '../types';

export type ControlType = { Status: number; MessageCode: number; Message: string };

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const prepareBody = <RequestBodyType extends Record<string, any>>(body: RequestBodyType, url: string) => {
  const requestBody = {
    Request: {
      Header: {
        DeviceId: '6cf35d17989d3fbe',
        RequestTime: new Date().getTime(),
        RequestId: uuidV1(),
        Version: v2ApiRoutes.includes(url) ? 2.0 : parseFloat(process.env.REACT_APP_API_VERSION!),
        Product: Number(process.env.REACT_APP_PRODUCT_VERSION)
      },
      Body: body || {}
    }
  };

  return needsEnvelope(url) ? { Data: requestBody } : requestBody;
};

type LogType = 'success' | 'error';
type LogInfo = {
  startTime: number;
  url: string;
  fetchOptions: any;
  response: any;
  json?: any;
  type: LogType;
};
const log = ({
  startTime, url, fetchOptions, response, json, type
}: LogInfo) => {
  if (process.env.NODE_ENV === 'development') { // Compiles this block out in non-dev builds
    const { body, ...otherOptions } = fetchOptions;
    console.groupCollapsed(
      `HTTP request: ${url} %c${Date.now() - startTime}ms %c${type}`,
      'color: grey;',
      `color: ${type === 'success' ? 'green' : 'red'}`
    );
    console.log('Request body:', JSON.stringify(JSON.parse(body), null, 2));
    if (json) console.log('Response json', JSON.stringify(json, null, 2));
    console.log('Full request:', { body, ...otherOptions });
    console.log('Full response:', response);
    console.groupEnd();
  }
};

const doesBodyContainError = ({ Control }: { Control: ControlType }) => (
  !Control || Control.Status === 0 || Control.MessageCode === 404
);

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type MakeApiCallOptions<RequestBodyType extends Record<string, any>> =
  Partial<Omit<RequestInit, 'body'>> & { body: RequestBodyType };

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default async <RequestBodyType extends Record<string, any>, ResponseType>(
  url: string,
  { headers, body, ...otherOptions }: MakeApiCallOptions<RequestBodyType> = { body: {} as RequestBodyType }
) => {
  const tokens = await getTokens();

  const startTime = Date.now();
  const fetchOptions = {
    method: 'POST',
    headers: {
      'X-API-Key': tokens.apiKey,
      'Content-Type': 'application/json',
      'X-REQUEST-ID': uuidV1(),
      'app-name': 'RJIL_JioKart',
      ...(needsAuthentication(url) && { Authorization: `Bearer ${tokens.accessToken}` }),
      ...headers
    },
    body: JSON.stringify(prepareBody<RequestBodyType>(body, url)),
    ...otherOptions
  };
  const response = await fetch(url, fetchOptions);

  if (!response.ok) {
    if (process.env.NODE_ENV === 'development') {
      log({
        startTime, url, fetchOptions, response, type: 'error'
      });
    }
    throw new HTTPError(response);
  }
  const json = await response.json();

  if (doesBodyContainError(json)) {
    if (process.env.NODE_ENV === 'development') {
      log({
        startTime, url, fetchOptions, response, json, type: 'error'
      });
    }
    throw new APIError(json);
  }

  if (process.env.NODE_ENV === 'development') {
    log({
      startTime, url, fetchOptions, response, json, type: 'success'
    });
  }

  return json.Data ? json.Data as ResponseType : json.Control as ResponseType;
};
