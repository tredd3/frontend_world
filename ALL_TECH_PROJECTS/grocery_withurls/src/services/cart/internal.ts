import promiseMemoize from 'promise-memoize';
import { apiRoutes } from '../api-routes';
import makeApiCall from '../helpers/make-api-call';
import { APICart, APIDeliveryOptions } from '../../types/api';
import { getUser } from '../user';
import { pivotCartResponse, pivotDeliveryMode } from '../helpers/pivots';
import { headerWithStoreId, actions } from './shared';
import { getStoreIdFromAddressId, getAddressById } from '../address';
import { DeliveryModesAPIRequestType, CheckoutRequest } from './types';

export const getCart = promiseMemoize(async (storeId?: number) => {
  const { userId, preferences: { storeId: defaultStoreId } } = await getUser();
  const selectedStoreId = storeId || defaultStoreId;

  const cartData: APICart = await makeApiCall<{}, APICart>(apiRoutes.getCart, {
    body: {
      userId,
      ...(selectedStoreId ? { storeId: selectedStoreId } : {})
    },
    ...headerWithStoreId(selectedStoreId)
  });

  return pivotCartResponse(cartData);
}, { resolve: 'json' });

export const clearInternalCart = () => (getCart.clear());

export function getUserIdStoreIdAndCartId (): Promise<{ userId: number; storeId?: number; cartId: number }>;
export function getUserIdStoreIdAndCartId (addressId: number): Promise<{ userId: number; storeId: number; cartId: number }>;
export async function getUserIdStoreIdAndCartId(addressId?: number) {
  // Don't make these calls in parallel, as getCart internally calls getUser.
  const { userId, preferences: { storeId } } = await getUser();
  const { cartId } = await getCart();

  if (!addressId) return { userId, storeId, cartId };

  const addressStoreId = await getStoreIdFromAddressId(addressId);
  if (!addressStoreId) throw new Error('Address doesn\'t have an associated store');

  return { userId, storeId: addressStoreId, cartId };
}

export const getDeliveryModes = promiseMemoize(async (addressId: number) => {
  const { totalPayableAmount } = await getCart();
  const address = await getAddressById(addressId);
  const storeId = await getStoreIdFromAddressId(addressId);
  const apiDeliveryOptions = await makeApiCall<DeliveryModesAPIRequestType, APIDeliveryOptions>(apiRoutes.getDeliveryModes, {
    body: {
      StoreId: Number(storeId),
      OrderValue: totalPayableAmount,
      Pincode: address!.pincode
    },
    ...headerWithStoreId(storeId)
  });

  return apiDeliveryOptions.StoreOrder.StorePickup.map(pivotDeliveryMode);
}, { resolve: 'json' });

export const deliveryDateByDateId = async (addressId: number, dateId: number) => {
  const deliveryModes = await getDeliveryModes(addressId);
  const matchingDate = deliveryModes?.find(date => date.id === dateId);
  if (!matchingDate) throw new Error('dateId is not valid');
  return matchingDate.value;
};

export const applyAddressChangeToCart = promiseMemoize(async (addressId: number) => {
  const storeId = await getStoreIdFromAddressId(addressId);
  if (!storeId) throw new Error('Address doesn\'t have an associated store');

  const oldCart = await getCart();
  const newCart = await getCart(storeId);

  return {
    cartHasBeenUpdated: oldCart.totalPayableAmount !== newCart.totalPayableAmount,
    oldCart,
    newCart
  };
}, { resolve: 'json' });

export const checkoutWithDeliveryDate = promiseMemoize(async (addressId: number, dateId: number) => {
  const { userId, storeId } = await getUserIdStoreIdAndCartId(addressId);
  const address = await getAddressById(addressId);
  const oldCart = await getCart();

  if (!address) throw new Error('addressId is not valid');

  const checkoutResponse = await makeApiCall<CheckoutRequest, APICart>(apiRoutes.checkout, {
    body: {
      storeId,
      cartId: oldCart.cartId,
      userId,
      customerAddressId: address.id,
      isAutoSelected: 0,
      deliveryDate: await deliveryDateByDateId(addressId, dateId),
      action: actions.postCheckout,
      timeslot: ''
    },
    ...headerWithStoreId(storeId)
  });

  const newCart = pivotCartResponse(checkoutResponse);

  return {
    cartHasBeenUpdated: oldCart.totalPayableAmount !== newCart.totalPayableAmount,
    oldCart,
    newCart
  };
}, { resolve: 'json' });

export const clearMemos = () => {
  getCart.clear();
  // eslint-disable-next-line @typescript-eslint/no-use-before-define, no-use-before-define
  checkoutWithDeliveryDate.clear();
  // eslint-disable-next-line @typescript-eslint/no-use-before-define, no-use-before-define
  applyAddressChangeToCart.clear();
};
