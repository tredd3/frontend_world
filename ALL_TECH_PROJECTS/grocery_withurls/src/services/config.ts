import promiseMemoize from 'promise-memoize';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { getUser } from './user';
import { APIConfig, ConfigPivot } from '../types';
import { pivotConfig } from './helpers/pivots';

export const getConfig = promiseMemoize(async (): Promise<ConfigPivot> => {
  const { userId } = await getUser();
  const { Configurations: apiConfig } = await makeApiCall<
    { customerId: number },
    {Configurations: APIConfig}
  >(apiRoutes.config, { body: { customerId: userId } });

  return pivotConfig(apiConfig);
});

export const deprecatedGetConfigDATA = async () => {
  try {
    const config = await getConfig();

    let LAST_ORDER_DETAILS = {};
    if (config.lastOrderDetails && Object.keys(config.lastOrderDetails).length !== 0) {
      LAST_ORDER_DETAILS = {
        TOTAL_ITEMS: config.lastOrderDetails.totalItems,
        ORDER_ID: config.lastOrderDetails.orderId,
        STORE_NAME: config.lastOrderDetails.storeName,
        TOTAL_SHIPMENT: config.lastOrderDetails.totalShipment,
        DELIVERED_DATE: config.lastOrderDetails.deliveredDate,
        STORE_ID: config.lastOrderDetails.storeName,
        ORDER_AMOUNT: config.lastOrderDetails.orderAmount,
        DELIVERED_TIME: config.lastOrderDetails.deliveredDate
      };
    }

    return {
      DEFAULT_PAGE_SIZE: config.pageSize,
      QTY_ERROR_MESSAGE: config.quantityErrorMessage,
      INVALID_OTP: config.invalidOTPMessage,
      MAX_COD_LIMIT: config.maxCODLimit,
      DELIVERY_CHARGE_THRESHOLD_AMOUNT: config.deliveryChargeThreshold,
      UNREAD_NOTIFICATION_COUNT: config.unreadNotificationCount,
      DELIVERY_CHARGES: config.deliveryCharges,
      COD_LIMIT_ERROR_MESSAGE: config.codLimitErrorMessage,
      WELCOME_PRODUCT_SKU: config.welcomeProductSKU,
      GENERIC_ERROR_MESSAGE: config.genericErrorMessage,
      MAX_QTY_LIMIT_PER_ITEM: config.maxQuantityLimitPerItem,
      LAST_ORDER_DETAILS
    };
  } catch (e) {
    return {};
  }
};
