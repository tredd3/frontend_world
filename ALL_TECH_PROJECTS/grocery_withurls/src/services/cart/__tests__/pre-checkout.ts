import { mocked } from 'ts-jest/utils';
import mockDB from '../../../mocks/db.json';
import * as internal from '../internal';
import * as user from '../../user';
import * as config from '../../config';
import {
  pivotCartResponse, pivotProduct,
  pivotConfig, pivotShipmentProduct,
  pivotDeliveryMode,
  pivotAPIUser
} from '../../helpers/pivots';
import { APIProduct, APIStorePickup } from '../../../types/api';
import * as makeApiCall from '../../helpers/make-api-call';
import {
  addToCart, moveToSaveForLater, moveToCart,
  removeFromSavedLater, changeQuantity,
  removeFromCart, getDeliveryDates, confirmChangedCart
} from '../pre-checkout';
import { apiRoutes } from '../../api-routes';

jest.mock('../../user');
jest.mock('../internal');
jest.mock('../../config');

const {
  getCart, getUserIdStoreIdAndCartId, clearMemos, clearInternalCart, getDeliveryModes
} = mocked(internal);

const {
  getUser
} = mocked(user);

const { getConfig } = mocked(config);

describe('pre-checkout', () => {
  let makeApiCallSpy: jest.SpyInstance<Promise<unknown>>;

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  clearMemos.mockImplementation(() => {});

  afterEach(() => {
    makeApiCallSpy.mockRestore();
  });

  describe('addToCart', () => {
    it.each([undefined, 1])(
      'should add an item to cart if it doesn\'t already exist',
      async arg => {
        getUserIdStoreIdAndCartId.mockImplementation(async () => ({
          userId: 111111, storeId: 222222, cartId: 333333
        }));
        getCart.mockImplementation(async () => pivotCartResponse(mockDB.cart.Data));
        getConfig.mockImplementation(async () => pivotConfig(mockDB.configurations.Data.Configurations));
        makeApiCallSpy = jest.spyOn(makeApiCall, 'default')
          .mockImplementationOnce(async () => mockDB.addItemToCart);

        await addToCart(pivotProduct(mockDB.product.Data.StoreProduct), arg);

        expect(makeApiCallSpy.mock.calls[0][0]).toBe(apiRoutes.addCartItem);
        expect(makeApiCallSpy.mock.calls[0][1]).toMatchSnapshot();
        expect(clearMemos).toHaveBeenCalled();
      }
    );

    it('should increase count if item already exists', async () => {
      getUserIdStoreIdAndCartId.mockImplementation(async () => ({
        userId: 111111, storeId: 222222, cartId: 333333
      }));
      getCart.mockImplementation(async () => pivotCartResponse(mockDB.cart.Data));
      getConfig.mockImplementation(async () => pivotConfig(mockDB.configurations.Data.Configurations));
      makeApiCallSpy = jest.spyOn(makeApiCall, 'default')
        .mockImplementationOnce(async () => mockDB.addItemToCart);

      await addToCart(
        pivotProduct(mockDB.cart.Data.Shipments[0].Product[0] as unknown as APIProduct),
        5
      );

      expect(makeApiCallSpy.mock.calls[0][0]).toBe(apiRoutes.addCartItem);
      expect(makeApiCallSpy.mock.calls[0][1]).toMatchSnapshot();
      expect(clearMemos).toHaveBeenCalled();
    });

    it('shouldn\'t allow adding new items than the max limit from config', async () => {
      getUserIdStoreIdAndCartId.mockImplementation(async () => ({
        userId: 111111, storeId: 222222, cartId: 333333
      }));
      getCart.mockImplementation(async () => pivotCartResponse(mockDB.cart.Data));
      getConfig.mockImplementation(async () => ({
        ...pivotConfig(mockDB.configurations.Data.Configurations),
        maxQuantityLimitPerItem: 5
      }));

      await expect(
        addToCart(pivotProduct(mockDB.product.Data.StoreProduct), 6)
      ).rejects.toBeInstanceOf(Error);
    });

    it('shouldn\'t allow adding to existing items than the max limit from config', async () => {
      getUserIdStoreIdAndCartId.mockImplementation(async () => ({
        userId: 111111, storeId: 222222, cartId: 333333
      }));
      const apiCart = JSON.parse(JSON.stringify(mockDB.cart.Data));
      apiCart.Shipments[0].Product[0].SelectedQuantity = 4;
      getCart.mockImplementation(async () => pivotCartResponse(apiCart));
      getConfig.mockImplementation(async () => ({
        ...pivotConfig(mockDB.configurations.Data.Configurations),
        maxQuantityLimitPerItem: 5
      }));

      await expect(addToCart(
        pivotProduct(mockDB.cart.Data.Shipments[0].Product[0] as unknown as APIProduct),
        2
      )).rejects.toBeInstanceOf(Error);
    });
  });

  describe('moveToSaveForLater', () => {
    it('should make the API call correctly', async () => {
      getUserIdStoreIdAndCartId.mockImplementation(async () => ({
        userId: 111111, storeId: 222222, cartId: 333333
      }));
      makeApiCallSpy = jest.spyOn(makeApiCall, 'default')
        .mockImplementation(async () => mockDB.saveForLater);

      await moveToSaveForLater(
        pivotShipmentProduct(mockDB.cart.Data.Shipments[0].Product[0])
      );

      expect(makeApiCallSpy.mock.calls[0][0]).toBe(apiRoutes.moveCartItem);
      expect(makeApiCallSpy.mock.calls[0][1]).toMatchSnapshot();
      expect(clearMemos).toHaveBeenCalled();
    });
  });

  describe('moveToCart', () => {
    it('should make the API call correctly', async () => {
      getUserIdStoreIdAndCartId.mockImplementation(async () => ({
        userId: 111111, storeId: 222222, cartId: 333333
      }));
      makeApiCallSpy = jest.spyOn(makeApiCall, 'default')
        .mockImplementation(async () => null);

      await moveToCart(
        pivotShipmentProduct(mockDB.cart.Data.Shipments[0].Product[0])
      );

      expect(makeApiCallSpy.mock.calls[0][0]).toBe(apiRoutes.moveCartItem);
      expect(makeApiCallSpy.mock.calls[0][1]).toMatchSnapshot();
      expect(clearMemos).toHaveBeenCalled();
    });
  });

  describe('removeFromSavedLater', () => {
    it('should make the API call correctly', async () => {
      clearInternalCart.mockImplementation(() => null);
      makeApiCallSpy = jest.spyOn(makeApiCall, 'default')
        .mockImplementation(async () => null);

      await removeFromSavedLater(
        pivotShipmentProduct(mockDB.cart.Data.Shipments[0].Product[0])
      );

      expect(makeApiCallSpy.mock.calls[0][0]).toBe(apiRoutes.removeCartSavedItem);
      expect(makeApiCallSpy.mock.calls[0][1]).toMatchSnapshot();
      expect(clearInternalCart).toHaveBeenCalled();
    });
  });

  describe('changeQuantity', () => {
    it('should change the quantity in cart as per the request', async () => {
      getUserIdStoreIdAndCartId.mockImplementation(async () => ({
        userId: 111111, storeId: 222222, cartId: 333333
      }));
      getConfig.mockImplementation(async () => pivotConfig(mockDB.configurations.Data.Configurations));
      makeApiCallSpy = jest.spyOn(makeApiCall, 'default')
        .mockImplementationOnce(async () => null);

      await changeQuantity(pivotShipmentProduct(mockDB.cart.Data.Shipments[0].Product[0]), 1);

      expect(makeApiCallSpy.mock.calls[0][0]).toBe(apiRoutes.addCartItem);
      expect(makeApiCallSpy.mock.calls[0][1]).toMatchSnapshot();
      expect(clearMemos).toHaveBeenCalled();
    });
  });

  it('should throw error if quantity is less than 1', async () => {
    getUserIdStoreIdAndCartId.mockImplementation(async () => ({
      userId: 111111, storeId: 222222, cartId: 333333
    }));
    getConfig.mockImplementation(async () => pivotConfig(mockDB.configurations.Data.Configurations));

    await expect(
      changeQuantity(pivotShipmentProduct(mockDB.cart.Data.Shipments[0].Product[0]), 0)
    ).rejects.toBeInstanceOf(Error);
  });

  it('should throw error if quantity is greater than max limit in config', async () => {
    getUserIdStoreIdAndCartId.mockImplementation(async () => ({
      userId: 111111, storeId: 222222, cartId: 333333
    }));
    getConfig.mockImplementation(async () => ({
      ...pivotConfig(mockDB.configurations.Data.Configurations),
      maxQuantityLimitPerItem: 5
    }));

    await expect(
      changeQuantity(pivotShipmentProduct(mockDB.cart.Data.Shipments[0].Product[0]), 6)
    ).rejects.toBeInstanceOf(Error);
  });

  describe('removeFromCart', () => {
    it('should make the API call correctly', async () => {
      getUserIdStoreIdAndCartId.mockImplementation(async () => ({
        userId: 111111, storeId: 222222, cartId: 333333
      }));
      makeApiCallSpy = jest.spyOn(makeApiCall, 'default')
        .mockImplementation(async () => mockDB.saveForLater);
      await removeFromCart(
        pivotShipmentProduct(mockDB.cart.Data.Shipments[0].Product[0])
      );

      expect(makeApiCallSpy.mock.calls[0][0]).toBe(apiRoutes.removeCartItem);
      expect(makeApiCallSpy.mock.calls[0][1]).toMatchSnapshot();
    });
  });

  describe('getDeliveryDates', () => {
    const addressId = 10160580;
    it('should return null if delivery modes not found', async () => {
      getDeliveryModes.mockImplementation(async () => []);
      const data = await getDeliveryDates(addressId);
      expect(data).toEqual([]);
      expect(getDeliveryModes).toHaveBeenCalledWith(addressId);
    });

    it('should return modes transformed as per the mapping function', async () => {
      getDeliveryModes.mockImplementation(
        async () => (mockDB.modes.Data.StoreOrder.StorePickup as APIStorePickup[]).map(pivotDeliveryMode)
      );
      const apiResp = await getDeliveryDates(addressId);
      expect(getDeliveryModes).toHaveBeenCalledWith(addressId);
      expect(apiResp).toMatchSnapshot();
    });
  });

  describe('confirmChangedCart', () => {
    it('should make the API call correctly', async () => {
      getUser.mockImplementation(async () => pivotAPIUser(mockDB.user.Data));

      makeApiCallSpy = jest.spyOn(makeApiCall, 'default')
        .mockImplementation(async () => mockDB.cart);

      await confirmChangedCart();
      expect(makeApiCallSpy.mock.calls[0][0]).toBe(apiRoutes.getCart);
      expect(makeApiCallSpy.mock.calls[0][1]).toMatchSnapshot();
    });
  });
});
