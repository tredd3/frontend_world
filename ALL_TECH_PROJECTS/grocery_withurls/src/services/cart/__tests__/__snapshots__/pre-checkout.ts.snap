// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pre-checkout addToCart should add an item to cart if it doesn't already exist 1`] = `
Object {
  "body": Object {
    "action": 0,
    "cartId": 333333,
    "product": Array [
      Object {
        "productType": 1,
        "qty": 1,
        "skuId": 10006440,
      },
    ],
    "storeId": 222222,
    "userId": 111111,
  },
  "headers": Object {
    "StoreId": "222222",
  },
}
`;

exports[`pre-checkout addToCart should add an item to cart if it doesn't already exist 2`] = `
Object {
  "body": Object {
    "action": 0,
    "cartId": 333333,
    "product": Array [
      Object {
        "productType": 1,
        "qty": 1,
        "skuId": 10006440,
      },
    ],
    "storeId": 222222,
    "userId": 111111,
  },
  "headers": Object {
    "StoreId": "222222",
  },
}
`;

exports[`pre-checkout addToCart should increase count if item already exists 1`] = `
Object {
  "body": Object {
    "action": 0,
    "cartId": 333333,
    "product": Array [
      Object {
        "productType": 1,
        "qty": 6,
        "skuId": 10028095,
      },
    ],
    "storeId": 222222,
    "userId": 111111,
  },
  "headers": Object {
    "StoreId": "222222",
  },
}
`;

exports[`pre-checkout changeQuantity should change the quantity in cart as per the request 1`] = `
Object {
  "body": Object {
    "action": 0,
    "cartId": 333333,
    "product": Array [
      Object {
        "productType": 1,
        "qty": 1,
        "skuId": 10028095,
      },
    ],
    "userId": 111111,
  },
  "headers": Object {
    "StoreId": "222222",
  },
}
`;

exports[`pre-checkout confirmChangedCart should make the API call correctly 1`] = `
Object {
  "body": Object {
    "action": 1,
    "userId": 10005564,
  },
}
`;

exports[`pre-checkout getDeliveryDates should return modes transformed as per the mapping function 1`] = `
Array [
  Object {
    "id": 2,
    "isOpen": true,
    "label": "Tomorrow - 07 January 2020",
  },
  Object {
    "id": 3,
    "isOpen": true,
    "label": "Wednesday - 08 January 2020",
  },
  Object {
    "id": 4,
    "isOpen": true,
    "label": "Thursday - 09 January 2020",
  },
]
`;

exports[`pre-checkout moveToCart should make the API call correctly 1`] = `
Object {
  "body": Object {
    "cartId": 333333,
    "moveType": "MoveToCart",
    "productType": 1,
    "qty": 1,
    "skuId": 10028095,
    "userId": 111111,
  },
  "headers": Object {
    "StoreId": "222222",
  },
}
`;

exports[`pre-checkout moveToSaveForLater should make the API call correctly 1`] = `
Object {
  "body": Object {
    "cartId": 333333,
    "moveType": "SaveForLater",
    "productType": 1,
    "qty": 1,
    "skuId": 10028095,
    "userId": 111111,
  },
  "headers": Object {
    "StoreId": "222222",
  },
}
`;

exports[`pre-checkout removeFromCart should make the API call correctly 1`] = `
Object {
  "body": Object {
    "action": 0,
    "cartId": 333333,
    "qty": 0,
    "skuId": 10028095,
    "type": 1,
    "userId": 111111,
  },
  "headers": Object {
    "StoreId": "222222",
  },
}
`;

exports[`pre-checkout removeFromSavedLater should make the API call correctly 1`] = `
Object {
  "body": Object {
    "productCategoryType": "SaveForLater",
    "productType": 1,
    "qty": 1,
    "skuId": 10028095,
  },
}
`;
